### PR TITLE
Extract interfaces and pull up optimizations from Skills Coach

### PR DESCRIPTION
**Pull request by williamboxhall**
**Best reviewed commit-by-commit**

This introduces:
* Interfaces for CommandGateway, EventProcessor and SequencedEventProcessor
* EventSequenceStats table to drastically improve performance of EventStore.lastSequence
* Other small bits and bobs from Skills Coach that we'd been waiting to lift up

### CODE REVIEW

**Review by wfoxd — APPROVED** [review]
Excellent, we can use the new processor from here.

File: src/main/kotlin/com/cultureamp/eventsourcing/BatchedAsyncEventProcessor.kt [hunk 1]
```diff
@@ -3,38 +3,61 @@ package com.cultureamp.eventsourcing
...
+        bookmarkStore: BookmarkStore,
+        bookmarkName: String,
+        eventProcessor: EventProcessor<M>,
+        batchSize: Int = 1000,
+        startLog: (Bookmark) -> Unit = { bookmark ->
+            System.out.println("Polling for events for ${bookmark.name} from sequence ${bookmark.sequence}")
+        },
+        endLog: (Int, Bookmark) -> Unit = { count, bookmark ->
+            if (count > 0 || Random.nextFloat() < 0.01) {
+                System.out.println("Finished processing batch for ${bookmark.name}, $count events up to sequence ${bookmark.sequence}")
```
[the remark this answers is not in the document]
Comment (adrianmancuso) [reply src/main/kotlin/com/cultureamp/eventsourcing/BatchedAsyncEventProcessor.kt:40]: Note, spoke to Will and it does.

File: src/test/kotlin/com/cultureamp/eventsourcing/RelationalDatabaseEventStoreTest.kt [hunk 2]
```diff
@@ -1,32 +1,40 @@
 package com.cultureamp.eventsourcing
 
-import com.cultureamp.eventsourcing.sample.*
+import com.cultureamp.eventsourcing.sample.PizzaCreated
+import com.cultureamp.eventsourcing.sample.PizzaEaten
 import com.cultureamp.eventsourcing.sample.PizzaStyle.MARGHERITA
-import com.cultureamp.eventsourcing.sample.PizzaTopping.*
+import com.cultureamp.eventsourcing.sample.PizzaTopping.BASIL
+import com.cultureamp.eventsourcing.sample.PizzaTopping.CHEESE
+import com.cultureamp.eventsourcing.sample.PizzaTopping.TOMATO_PASTE
```
Comment (adrianmancuso) [inline src/test/kotlin/com/cultureamp/eventsourcing/RelationalDatabaseEventStoreTest.kt:8]: 👨🏽‍🍳

File: src/test/kotlin/com/cultureamp/eventsourcing/AsyncEventProcessorMonitorIntegrationTest.kt [hunk 3]
```diff
@@ -44,13 +45,15 @@ class AsyncEventProcessorMonitorIntegrationTest : DescribeSpec({
         transaction(db) {
             addLogger(StdOutSqlLogger)
             SchemaUtils.create(table)
+            SchemaUtils.create(eventsSequenceStatsTable)
```
Comment (adrianmancuso) [inline src/test/kotlin/com/cultureamp/eventsourcing/AsyncEventProcessorMonitorIntegrationTest.kt:48]: One of these days we need to figure out when this is necessary and when it's a random incantation that we do for good luck. Did anything fail before you added this?

**Review by adrianmancuso — APPROVED** [review]

File: src/test/kotlin/com/cultureamp/eventsourcing/AsyncEventProcessorMonitorIntegrationTest.kt (hunk 3 above)
Comment (williamboxhall) [reply src/test/kotlin/com/cultureamp/eventsourcing/AsyncEventProcessorMonitorIntegrationTest.kt:48]: Hahaha nah in this case it definitely failed without this, saying the table couldn't be found

**Pull request merged**